Add documentation of just_once (#234)
SFDO-Tooling/Snowfakery: docs/index.md

## Advanced Features

### Singletons with the "just_once" feature

Snowfakery scales up to larger data volumes
by evaluating your recipe over and over again.
Each one is called an `iteration`.

Some objects are meant to be produced only once, regardless
of how many times the recipe executes. The programming

> language term for thhis is a "singleton". For example

Author: ```suggestion
language term for this is a "singleton". For example
```

> an accounting system might generate a datase that has

Author: ```suggestion
an accounting system might generate a dataset that has
```

> exactly 3 Ledger objects, Checking and Savings. You could have

Reviewer: Do you mean 2 here? Or is it 3 Ledger object + 1 checking and 1 savings?
```suggestion
exactly 2 Ledger objects, Checking and Savings. You could have
```

dozens of Entries per Ledger or maybe billions. But it

> might want to always generate exactly 3 Ledgers.

Reviewer: same question here.

> might want to always generate exactly 3 Ledgers.

Author: ```suggestion
might want to always generate exactly 2 Ledgers.
```

Here is how you would do that:

```yaml
- object: Ledger
  just_once: True
  nickname: Checking
  fields:
    Name: Checking

- object: Ledger
  just_once: True
  nickname: Savings
  fields:
    Name: Savings

- object: Entry
  count: 1000
  fields:
    Ledger: Checking
    ...

- object: Entry
  count: 1000
  fields:
    Ledger: Savings
    ...
```

Now if you execute this from the Snowfakery command line like this:

> ```s

Reviewer: I don't think the 's' is needed.

Author: It seems to format the text as a shell script. I guess I could have called it "sh" to be clearer.

Author: Experimentally "s" works best.

    ```s 

```s 
$ foo bar baz 
xyzzy
```

```bash
$ foo bar baz 
xyzzy
```

```sh
$ foo bar baz 
xyzzy
```

works.

Reviewer: TIL

> ```s
> $ snowfakery accounting.yml --target-number 10_000 Entry
> ...
> ```

Reviewer: ```suggestion
```
$ snowfakery accounting.yml --target-number 10_000 Entry
...
```
```

- Object Template: These represent instructions on how to create a row, or multiple rows in a database. Each row represents a real-world Object.
- Rows: Rows (often also called “records”) in a database are a unit of related information. For example in Salesforce (which includes a database) a “Contact” has a first name, last name, phone number, etc. Each Contact is a row. “Contact” is the type of each of those rows. Rows represent real-world Objects. See “Objects” above for more information.
- Recipe: A Snowfakery YAML file instructing Snowfakery on what to generate.
- Iteration: Snowfakery recipes can be "scaled up" to generate more data by specifying command line, API or CumulusCI options. The recipe is scaled up by executing over and over again. These executions are called iterations.

> - Singleton: A singleton is an Object Template that generates a single row now how many times the recipe is iterated over.

Reviewer: ```suggestion
- Singleton: A singleton is an Object Template that generates a single row regardless of how many times the recipe is iterated over.
```
